src/cunumeric/stat/histogram_impl.h

```diff
@@ -106,5 +114,66 @@ void histogram_weights(exe_policy_t exe_pol,
   segsum();
 }
 
+template <typename exe_policy_t, typename elem_t, typename bin_t, typename weight_t>
+void histogram_wrapper(exe_policy_t exe_pol,
+                       const AccessorRO<elem_t, 1>& src,
+                       const Rect<1>& src_rect,
+                       const AccessorRO<bin_t, 1>& bins,
+                       const Rect<1>& bins_rect,
+                       const AccessorRO<weight_t, 1>& weights,
+                       const Rect<1>& weights_rect,
+                       const AccessorRD<SumReduction<weight_t>, true, 1>& result,
+                       const Rect<1>& result_rect,
+                       cudaStream_t stream = nullptr)
+{
+  auto&& [src_size, src_copy, src_ptr] = accessors::make_accessor_copy(exe_pol, src, src_rect);
+
+  auto&& [weights_size, weights_copy, weights_ptr] =
+    accessors::make_accessor_copy(exe_pol, weights, weights_rect);
+
+  assert(weights_size == src_size);
+
+  auto&& [bins_size, bins_ptr] = accessors::get_accessor_ptr(bins, bins_rect);
+
+  auto num_intervals            = bins_size - 1;
+  Buffer<weight_t> local_result = create_buffer<weight_t>(num_intervals);
+
+  weight_t* local_result_ptr = local_result.ptr(0);
+
+  auto&& [global_result_size, global_result_ptr] = accessors::get_accessor_ptr(result, result_rect);
+
+#ifdef CUDA_STREAM_CHECKER
```

Comment:
So, `if constexpr(...)` cannot be a substitute for a pre-processor macro, like `#ifdef`. That's because the compile-time `false` path of an `if constexpr(...)` still gets _compiled_ .

For example (see https://en.cppreference.com/w/cpp/language/if):
```
void f()
{
    if constexpr(false)
    {
        int i = 0;
        int *p = i; // Error even though in discarded statement
    }
}
```

Reply:
The only option to `#ifdef CUDA_STREAM_CHECKER` is specializing a function template 
```
template <typename exe_policy_t> synchronize(exe_policy_t exe_pol){...}
```
The problem is function template cannot be _partially_ specialized. They need to be fully specialized. And neither the host nor the gpu policies are unique. That means having to explicitly specialize on every possible thrust execution policy.

But, I can provide a template class to that effect, which _can_ be partially specialized.

Reply:
> the compile-time false path of an if constexpr(...) still gets compiled .

Would `histogram_wrapper` fall into the exception mentioned later in https://en.cppreference.com/w/cpp/language/if?

> If a constexpr if statement appears inside a [templated entity](https://en.cppreference.com/w/cpp/language/templates#Templated_entity), and if condition is not [value-dependent](https://en.cppreference.com/w/cpp/language/dependent_name#Value-dependent_expressions) after instantiation, the discarded statement is not instantiated when the enclosing template is instantiated.

I think it does, by virtue of being a function template.

Reply:
Seems to work in one compiler at least https://godbolt.org/z/v6b4veGrz

Reply:
`histogram.cc`, and `histogram_omp.cc` fail to compile for me.

Reply:
```
In file included from /home/aschaffer/Development/Projects/cuNumeric/forked/fea_ext_histogram/histogram_refactor/cunumeric/src/cunumeric/stat/histogram.cc:21:
    /home/aschaffer/Development/Projects/cuNumeric/forked/fea_ext_histogram/histogram_refactor/cunumeric/src/cunumeric/stat/histogram_impl.h: In function 'void cunumeric::detail::histogram_wrapper(exe_policy_t, legate::AccessorRO<elem_t, 1>&, legate::Rect<1>&, legate::AccessorRO<bin_t, 1>&, legate::Rect<1>&, legate::AccessorRO<weight_t, 1>&, legate::Rect<1>&, legate::AccessorRD<Legion::SumReduction<weight_t>, true, 1>&, legate::Rect<1>&, cudaStream_t)':
    /home/aschaffer/Development/Projects/cuNumeric/forked/fea_ext_histogram/histogram_refactor/cunumeric/src/cunumeric/stat/histogram_impl.h:145:52: error: there are no arguments to 'CHECK_CUDA_STREAM' that depend on a template parameter, so a declaration of 'CHECK_CUDA_STREAM' must be available [-fpermissive]
      145 |   if constexpr (!is_host_policy_v<exe_policy_t>) { CHECK_CUDA_STREAM(stream); }
          |                                                    ^~~~~~~~~~~~~~~~~
    /home/aschaffer/Development/Projects/cuNumeric/forked/fea_ext_histogram/histogram_refactor/cunumeric/src/cunumeric/stat/histogram_impl.h:145:52: note: (if you use '-fpermissive', G++ will accept your code, but allowing the use of an undeclared name is deprecated)
    /home/aschaffer/Development/Projects/cuNumeric/forked/fea_ext_histogram/histogram_refactor/cunumeric/src/cunumeric/stat/histogram_impl.h:156:52: error: there are no arguments to 'CHECK_CUDA_STREAM' that depend on a template parameter, so a declaration of 'CHECK_CUDA_STREAM' must be available [-fpermissive]
      156 |   if constexpr (!is_host_policy_v<exe_policy_t>) { CHECK_CUDA_STREAM(stream); }
```

Reply:
Sounds like I was giving `if constexpr` too much credit.

> But, I can provide a template class to that effect, which can be partially specialized.

Would this still work?

Reply:
Yes, I can use a `sync_policy_t` which then can be specialized and used similarly to `segmented_sum_t`.

Reply:
Ok, done.